### PR TITLE
thrift_proxy: introduce header transport

### PR DESCRIPTION
**Pull request by zuercher**
Implements the Thrift "header" transport which primarily provides
key/value pairs that may contain arbitrary message metadata.

*Risk Level*: low
*Testing*: added unit/integration tests
*Docs Changes*: n/a
*Release Notes*: added note about header transport support

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

### CODE REVIEW

**Comment by zuercher** [comment]
@rgs1 or @fishcakez github won't let me directly request reviews or assign you, but if you wanted to take a look at this, I'd be grateful.

**Comment by rgs1** [comment]
@zuercher thanks for tagging us, I"ll look today.

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    break;
+  case HeaderProtocolType::Compact:
+    proto = ProtocolType::Compact;
+    break;
+  default:
+    throw EnvoyException(fmt::format("Unknown protocol {}", static_cast<int>(header_proto)));
+  }
+  metadata.setProtocol(proto);
+
+  int16_t num_xforms = drainVarIntI16(buffer, header_size, "transform count");
```
Comment (brian-pane) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: Is this line vulnerable to a buffer read overflow? Here's the scenario I'm worried about:
* A malicious client sends a 15-byte message where `header_size` = 1
* This message gets through all the checks on lines 57-66 above.
* Then line 72 removes the first 14 bytes, leaving 1 byte in `buffer`.
* Depending on the value in that 1 remaining byte, `drainVarInt16` may try to read the byte after it, which is beyond the end of `buffer`.

Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: I don't believe so. We require that the full header (as given by header size) by available. As we read bytes from the header we decrement header_size (which is passed as a reference to drainVarIntI16). drainVarIntI16 is just drainVarIntI32 with an upper bound check. The latter requires header_size to be at least 1 and then invokes BufferHelper::peekVarIntI32 which validates the buffer has enough data before reading each byte that composes the value. If the peek reports a buffer underflow (buffer doesn't have enough data to read the whole number) or requires more bytes than are left in the header (supposing the buffer has more data beyond the end of the header) we throw.

That said, I'll add a test that exercises this directly. Mostly as a remdinder to myself: It's basically [this one](https://github.com/envoyproxy/envoy/pull/4082/files#diff-464b7abb89897074c8d6d7964d2cddd3R160) except that the number of transforms should be  truncated.

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  Compact = 2,
+
+  FirstHeaderProtocolType = Binary,
+  LastHeaderProtocolType = Compact,
+};
+
+} // namespace
+
+bool HeaderTransportImpl::decodeFrameStart(Buffer::Instance& buffer, MessageMetadata& metadata) {
+  if (buffer.length() < 14) {
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: nit: should this be a constant (e.g.: `HEADER_LEN`)?

File: test/extensions/filters/network/thrift_proxy/config_test.cc [hunk 3]
```diff
@@ -14,38 +14,107 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace ThriftProxy {
 
-TEST(ThriftFilterConfigTest, ValidateFail) {
-  NiceMock<Server::Configuration::MockFactoryContext> context;
-  EXPECT_THROW(ThriftProxyFilterConfigFactory().createFilterFactoryFromProto(
-                   envoy::config::filter::network::thrift_proxy::v2alpha1::ThriftProxy(), context),
+namespace {
+
+std::vector<envoy::config::filter::network::thrift_proxy::v2alpha1::ThriftProxy_TransportType>
```
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/config_test.cc]: nit: should this be typedef'd (e.g.: `TransportTypeVector`)?

File: test/extensions/filters/network/thrift_proxy/config_test.cc [hunk 4]
```diff
@@ -14,38 +14,107 @@ namespace Extensions {
...
+                          ThriftProxy_TransportType_TransportType_MAX) {
+    v.push_back(static_cast<
+                envoy::config::filter::network::thrift_proxy::v2alpha1::ThriftProxy_TransportType>(
+        transport));
+    transport++;
+  }
+  return v;
+}
+
+std::vector<envoy::config::filter::network::thrift_proxy::v2alpha1::ThriftProxy_ProtocolType>
```
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/config_test.cc]: ditto

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc (hunk 2 above)
Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: Added some constants for the various sizes this transport uses.

**Comment by rgs1** [comment]
Thanks for the updates. LGTM.

**Comment by zuercher** [comment]
@mattklein123 can I get you to take a quick pass at this next time you're reviewing PRs?

**Comment by mattklein123** [comment]
@zuercher will do Thursday morning unless someone else gets to it sooner. Sorry for the delay.

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+  FirstHeaderProtocolType = Binary,
+  LastHeaderProtocolType = Compact,
+};
+
+// Fixed portion of frame header:
+//   Header magic: 2 bytes +
+//   Flags: 2 bytes +
+//   Sequence number: 4 bytes
+//   Header data size: 2 bytes
+constexpr uint64_t MIN_FRAME_START_SIZE_NO_HEADERS = 10;
```
Comment (mattklein123) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: nit: I think we are doing CamelCase now for constants. `MinFrameStartSizeNoHeaders`

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * HeaderTransportImpl implements the Thrift Header transport.
+ * See https://github.com/apache/thrift/blob/master/doc/specs/HeaderFormat.md and
+ * https://github.com/apache/thrift/blob/master/lib/cpp/src/thrift/transport/THeaderTransport.h
+ * (for constants not specified in the spec).
+ */
+class HeaderTransportImpl : public Transport {
+public:
+  HeaderTransportImpl() {}
```
Comment (mattklein123) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.h]: nit: Not needed

**Comment by zuercher** [comment]
@mattklein123 addressed your nits, could you restamp this when you have a moment?

**Pull request merged**